2601 appkey

hyi opened the pull request:
<!--

Please read, and add your text at the bottom of this message.

Thanks for contributing code to HydroShare. In order to maintain code quality and expedite this process, please assist the development team by making sure the following is present in this pull request.

For more information, see https://docs.google.com/document/d/1dzxqlZW5fKNEyQSeKiSFq-SmS-VOPCva95XXkBjPExs

-->

### Pull Request Checklist: 
- [x] Positive Test Case Written by Dev

<!-- Enter steps that a QA engineer, stakeholder, or user documentation writer would follow to test the positive or "successful" case of the functionality your code provides or fixes -->

- [x] Automated Testing

<!-- Our Jenkins Instance is set up to automatically test every commit from a pull request. Code coverage must not decrease so new functionality or code paths added during a bug fix must have appropriate tests written. Every test must pass, including PEP8 code formatting tests. -->

- [x] Sufficient User and Developer Documentation

<!-- Please email your positive test case lbrazil@cuahsi.org, who will make the decision regarding user documentation. -->

- [x] Passing Jenkins Build

<!-- Our Jenkins Instance is set up to automatically test every commit from a pull request. Every test must pass, including PEP8 code formatting tests. -->

- [x] Peer Code review and approval

<!-- This is the process by which a peer developer on the HydroShare team will read the changeset, provide feedback, and ultimately give a formal approval to the code before it passes PR status. -->

### Positive Test Case
1. [Enter positive test case here]
(1) create a web app tool resource and add extended metadata name-value pair with name as ```appkey``` and value as an application-specific value, say, ```model-my-watershed```.
(2) create a test resource upon which the web app tool created above can act, and add the same extended metadata name-value pair with name as ```appkey``` as shown above.
(3) With matching ```appkey``` name-value pair extended metadata, the web app tool resource automatically shows up in "Open with" list for the test resource as well as on the web app tab. 
This functionality is requested by MMW team and add another granularity of control for users to associate a resource with a specific web app tool resource.

hyi commented:
@zhiyuli I assigned you to review this PR since the code is all in web app tool resource implementation. @dtarb asked me to get this functionality implemented and approved as early as I can, so hopefully you can review this PR soon. Thanks.

hyi commented:
@zhiyuli @dtarb I deployed this branch on playground, so feel free to test it out as needed.

pkdash commented on hs_tools_resource/app_launch_helper.py:
```diff
@@ -9,39 +9,54 @@ def resource_level_tool_urls(resource_obj, request_obj):
+    # associate resources with app tools using extended metadata name-value pair with 'appkey' key
+    tool_app_key = 'appkey'
+    tool_kv_dict = {}
+    for tool_res_obj in ToolResource.objects.all():
```
@hyi I think you can simplify this code using the hstore field query api. Have a look at the documentation: https://docs.djangoproject.com/en/1.8/ref/contrib/postgres/fields/#querying-hstorefield

The code needed to find a list of web apps associated with a given resource can be done with only couple of lines of code once you use the hstore specific query api.

hyi replied:
@pkdash Thanks for the info. Will update the code to use the hstore field query api.

hydrobuild commented:
Build finished.

hyi commented:
@pkdash Have updated code to use hstorefield query API which indeed simplified code logic quite a bit. Thanks again for the info. Feel free to take a look and let me know if you see any further issues. @zhiyuli It is ready for your code review now that I have addressed @pkdash's comment. I have also updated playground to make sure the same query logic applies. Let me know if you have questions or notice any issues. Thanks.

hydrobuild commented:
Build finished.

hydrobuild commented:
Build finished.

hyi commented:
@zhiyuli Can you formally approve this PR so that green check shows up for ```Review required```? Even though you approved it, it is still blocked. So you may need to go through github approve PR UI to formally approve it.
